opm/flowdiagnostics/DerivedQuantities.cpp

```diff
@@ -0,0 +1,125 @@
+/*
+  Copyright 2015, 2016 SINTEF ICT, Applied Mathematics.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#if HAVE_CONFIG_H
+#include <config.h>
+#endif // HAVE_CONFIG_H
+
+#include <opm/flowdiagnostics/DerivedQuantities.hpp>
+#include <numeric>
+
+namespace Opm
+{
+namespace FlowDiagnostics
+{
+
+    /// The F-Phi curve.
+    ///
+    /// The F-Phi curve is an analogue to the fractional flow
+    /// curve in a 1D displacement. It can be used to compute
+    /// other interesting diagnostic quantities such as the Lorenz
+    /// coefficient. For a technical description see Shavali et
+    /// al. (SPE 146446), Shook and Mitchell (SPE 124625).
+    Graph flowCapacityStorageCapacityCurve(const Toolbox::Forward& injector_solution,
+                                           const Toolbox::Reverse& producer_solution,
+                                           const std::vector<double>& pv)
+    {
+        const auto& ftof = injector_solution.fd.timeOfFlight();
+        const auto& rtof = producer_solution.fd.timeOfFlight();
+        if (pv.size() != ftof.size() || pv.size() != rtof.size()) {
+            throw std::runtime_error("computeFandPhi(): Input vectors must have same size.");
+        }
+
+        // Sort according to total travel time.
+        const int n = pv.size();
+        typedef std::pair<double, double> D2;
+        std::vector<D2> time_and_pv(n);
+        for (int ii = 0; ii < n; ++ii) {
+            time_and_pv[ii].first = ftof[ii] + rtof[ii]; // Total travel time.
+            time_and_pv[ii].second = pv[ii];
+        }
+        std::sort(time_and_pv.begin(), time_and_pv.end());
```

Comment:
I **suppose** you could write this as
```C++
using Index = // Some index type
auto ix = std::vector<Index>(n);
std::iota(ix.begin(), ix.end(), Index(0))

std::sort(ix.begin(), ix.end(),
    [&ftof, &rtof, &pv](const Index i1, const Index i2) -> bool
{
    const auto p1 = std::make_pair(ftof[i1] + rtof[i1], pv[i1]);
    const auto p2 = std::make_pair(ftof[i2] + rtof[i2], pv[i2]);

    return p1 < p2;
});
```
and then work off of the sorted indices in the remaining code.  That would make the sort elements smaller and you would not need to carry around copies of the time-of-flight and pore-volume vectors.

Whether or not that actually makes a difference is another matter entirely of course.